utils.py

```diff
@@ -253,3 +315,222 @@ def availability_to_dict(availability_list):
+
+    _validate_availability(params, result, employees, errors, msgs)
+    _validate_shifts_per_employee(params, result, employees, errors, msgs)
+    _validate_employees_per_shift(params, result, errors, msgs)
+    _validate_max_consecutive_shifts(params, result, employees, errors, msgs)
+    _validate_trainee_shifts(params, result, employees, errors, msgs)
+    if params.requires_manager:
+        _validate_requires_manager(params, result, employees, errors, msgs)
+    if not params.allow_isolated_days_off:
+        _validate_isolated_days_off(params, result, employees, errors, msgs)
+
+    return errors
+
+
+def _validate_availability(
+    params: ModelParams,
+    results: np.ndarray,
+    employees: list[str],
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates employee availability for the solution and updates the `errors`
+    dictionary with any errors found. Requires the `msgs` dict to have the
+    key `'unavailable'`."""
+    msg_key, msg_template = msgs["unavailable"]
+    for e, employee in enumerate(employees):
+        for s, day in enumerate(params.shift_labels):
+            if results[e, s] > params.availability[employee][s]:
+                errors[msg_key].append(
+                    msg_template.format(employee=employee, day=day)
+                )
+    return errors
+
+
+def _validate_shifts_per_employee(
+    params: ModelParams,
+    results: np.ndarray,
+    employees: list[str],
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates the number of shifts per employee for the solution and updates
+    the `errors` dictionary with any errors found. Requires the `msgs` dict
+    to have the keys `'insufficient'` and `'overtime'`."""
+    insufficient_key, insufficient_template = msgs["insufficient"]
+    overtime_key, overtime_template = msgs["overtime"]
+    for e, employee in enumerate(employees):
+        num_shifts = results[e, :].sum()
+        if num_shifts < params.min_shifts:
+            errors[insufficient_key].append(
+                insufficient_template.format(employee=employee)
+            )
+        elif num_shifts > params.max_shifts:
+            errors[overtime_key].append(overtime_template.format(employee=employee))
+    return errors
+
+
+def _validate_employees_per_shift(
+    params: ModelParams,
+    results: np.ndarray,
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates the number of employees per shift for the solution and updates
+    the `errors` dictionary with any errors found. Requires the `msgs` dict
+    to have the keys `'understaffed'` and `'overstaffed'`."""
+    for s, day in enumerate(params.shift_labels):
+        understaffed_key, understaffed_template = msgs["understaffed"]
+        overstaffed_key, overstaffed_template = msgs["overstaffed"]
```

Comment:
I don't think these need to be assigned inside the loop

```suggestion
    understaffed_key, understaffed_template = msgs["understaffed"]
    overstaffed_key, overstaffed_template = msgs["overstaffed"]

    for s, day in enumerate(params.shift_labels):
```